tower-http/src/follow_redirect/mod.rs

```diff
@@ -0,0 +1,325 @@
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        let mut service = self.inner.clone();
+        let mut policy = self.policy.clone();
+        let mut body = BodyRepr::None;
+        body.try_clone_from(req.body(), &policy);
+        policy.on_request(&mut req);
+        ResponseFuture {
+            method: req.method().clone(),
+            uri: req.uri().clone(),
+            version: req.version(),
+            headers: req.headers().clone(),
+            body,
+            future: service.call(req),
+            service,
+            policy,
+        }
+    }
+}
+
+/// Response future for [`FollowRedirect`].
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, S, B, P> {
+    #[pin]
+    future: F,
+    service: S,
+    policy: P,
+    method: Method,
+    uri: Uri,
+    version: Version,
+    headers: HeaderMap<HeaderValue>,
+    body: BodyRepr<B>,
+}
+
+impl<F, S, ReqBody, ResBody, P> Future for ResponseFuture<F, S, ReqBody, P>
+where
+    F: Future<Output = Result<Response<ResBody>, S::Error>>,
+    S: Service<Request<ReqBody>, Future = F> + Clone,
+    ReqBody: Body + Default,
+    P: Policy<ReqBody, S::Error>,
+{
+    type Output = Result<Response<ResBody>, S::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let res = ready!(this.future.as_mut().poll(cx)?);
+
+        match res.status() {
+            StatusCode::MOVED_PERMANENTLY | StatusCode::FOUND => {
+                // User agents MAY change the request method from POST to GET
+                // (RFC 7231 section 6.4.2. and 6.4.3.).
+                if *this.method == Method::POST {
+                    *this.method = Method::GET;
+                    *this.body = BodyRepr::Empty;
+                }
+            }
+            StatusCode::SEE_OTHER => {
+                // A user agent can perform a GET or HEAD request (RFC 7231 section 6.4.4.).
+                if *this.method != Method::HEAD {
+                    *this.method = Method::GET;
+                }
+                *this.body = BodyRepr::Empty;
+            }
+            StatusCode::TEMPORARY_REDIRECT | StatusCode::PERMANENT_REDIRECT => {}
+            _ => return Poll::Ready(Ok(res)),
+        };
+
+        let body = if let Some(body) = this.body.take() {
+            body
+        } else {
+            return Poll::Ready(Ok(res));
+        };
+
+        let location = res
+            .headers()
+            .get(&LOCATION)
+            .and_then(|loc| str::from_utf8(loc.as_bytes()).ok())
+            .and_then(|loc| RiReferenceStr::<UriSpec>::new(loc).ok())
+            .and_then(|loc| {
+                Some(loc.resolve_against(&RiAbsoluteString::try_from(this.uri.to_string()).ok()?))
+            })
+            .and_then(|loc| Uri::try_from(loc.as_str()).ok());
```

Comment:
What does this actually do? I'm not familiar with [iri_string](https://docs.rs/iri-string/0.3.0/iri_string/index.html).

What do you think about extracting this bit into a function that returns `Option<_>` so we could use `?` without chaining `and_then` so many times?

Reply:
This is resolving the URI reference of the `Location` header against the request URI ([RFC 7231 section 7.1.2](https://tools.ietf.org/html/rfc7231#section-7.1.2)). I added an explanation to the extracted function.